### PR TITLE
DEPR: Add FutureWarning for pandas.io.sql.execute

### LINKED ISSUE #50185: DEPR: pandas.io.sql.execute
There is currently a bug in `pandas.io.sql.execute` on the main branch (not in any released versions). I created the bug in #49531 while working on  #48576. In #49531, I changed `SQLDatabase` to only accept a `Connection` and not an `Engine`, because sqlalchemy 2.0 restricts the methods that are available to `Engine`. #49531 created bugs in both `read_sql` and `pandas.io.sql.execute`, and I have an open PR to fix `read_sql` in #49967.

This reproduces the bug in `pandas.io.sql.execute`:
```
from pathlib import Path
from sqlalchemy import create_engine
from pandas import DataFrame
from pandas.io import sql

db_file = Path(r"C:\Temp\test.db")
db_file.unlink(missing_ok=True)
db_uri = "sqlite:///" + str(db_file)
engine = create_engine(db_uri)
DataFrame({'a': [2, 4], 'b': [3, 6]}).to_sql('test_table', engine)
sql.execute('select * from test_table', engine).fetchall()
```

Traceback when running on the main branch in pandas:
```
Error closing cursor
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\cursor.py", line 991, in fetchall
    rows = dbapi_cursor.fetchall()
sqlite3.ProgrammingError: Cannot operate on a closed database.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 1995, in _safe_close_cursor
    cursor.close()
sqlite3.ProgrammingError: Cannot operate on a closed database.
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\cursor.py", line 991, in fetchall
    rows = dbapi_cursor.fetchall()
sqlite3.ProgrammingError: Cannot operate on a closed database.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Github\pandas\execute_error.py", line 11, in <module>
    sql.execute('select * from test_table', engine).fetchall()
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\result.py", line 1072, in fetchall
    return self._allrows()
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\result.py", line 401, in _allrows
    rows = self._fetchall_impl()
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\cursor.py", line 1819, in _fetchall_impl
    return self.cursor_strategy.fetchall(self, self.cursor)
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\cursor.py", line 995, in fetchall
    self.handle_exception(result, dbapi_cursor, e)
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\cursor.py", line 955, in handle_exception
    result.connection._handle_dbapi_exception(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\util\compat.py", line 211, in raise_
    raise exception
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\cursor.py", line 991, in fetchall
    rows = dbapi_cursor.fetchall()
sqlalchemy.exc.ProgrammingError: (sqlite3.ProgrammingError) Cannot operate on a closed database.
(Background on this error at: https://sqlalche.me/e/14/f405)
```

When running on pandas 1.5.2, there is no error.

After #49967, `SQLDatabase` accepts `str`, `Engine`, and `Connection`, but in `SQLDatabase.__init__`, a `Connection` is created if it was not passed in, and an `ExitStack` is created to clean up the `Connection` and `Engine` if necessary. Cleanup happens either in `SQLDatabase.__exit__` or at the end of the generator which is returned by `read_sql` if `chunksize` is not None. What I'm not able to see is how to do the cleanup after the caller fetches the results of `pandas.io.sql.execute`. I have come up with two options so far:
* Restrict `pandas.io.sql.execute` to require a `Connection`, or else do not allow queries that return rows if a `str` or `Engine` is used.
* Use `Result.freeze` to fetch the results before closing the `SQLDatabase` context manager.  The problem here is that it requires sqlalchemy 1.4.45, which was only released on 12/10/2022, due to a bug in prior versions: https://github.com/sqlalchemy/sqlalchemy/issues/8963

### PR DESCRIPTION
**Pull request by luke396**
- [x] closes #50185
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/io/test_sql.py [hunk 1]
```diff
@@ -668,6 +668,15 @@ def test_execute_typeerror(sqlite_iris_engine):
         sql.execute("select * from iris", sqlite_iris_engine)
 
 
+def test_execute_deprecated(sqlite_buildin_iris):
+    with tm.assert_produces_warning(
```
Comment (phofl) [inline pandas/tests/io/test_sql.py:672]: Please add gh ref

**Review by phofl — COMMENTED** [review]
Doc build is failing, mightt be related.

Could you remove all mentions in the docs?

**Comment by luke396** [comment]
> Doc build is failing, mightt be related.
> 
> Could you remove all mentions in the docs?

Your mean is remove '`:func:`pandas.io.sql.execute` 'in .rst ?

**Comment by phofl** [comment]
Ah no, I think we mention this function in the user guide, this reference should be removed

**Comment by luke396** [comment]
> Ah no, I think we mention this function in the user guide, this reference should be removed

Fine, I will do it.

**Comment by luke396** [comment]
image

May be this need some attention, in check `Doc Build and Upload / Doc Build and Upload`

File: pandas/core/generic.py [hunk 2]
```diff
@@ -2858,8 +2858,11 @@ def to_sql(
 
         >>> df.to_sql('users', con=engine)
         3
-        >>> engine.execute("SELECT * FROM users").fetchall()
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/generic.py:2861]: Same for all the examples below

[the remark this answers is not in the document]
Comment (luke396) [reply pandas/core/generic.py]: > Same for all the examples below

I seem to mess up the whole PR, I will reopen a new on 😢

**Pull request closed without merging**